### DISCUSSION TITLE
Set-output Truncates Multiline Strings

### DISCUSSION THREAD

**Discussion by jacderida** [community/community#26288] [Actions]
<p>Hi,</p>
<p>I am using this code to create a release for one of our repositories:</p>
<pre><code>- shell: bash
        id: release_description
        run: |
          description=$(./resources/get_release_description.sh ${{ steps.versioning.outputs.cli_version }})
          echo $description
          echo "::set-output name=description::$description"
        if: startsWith(steps.commit_message.outputs.commit_message, 'Version change')
      - shell: bash
        run: |
          echo ${{ steps.release_description.outputs.description }}
        if: startsWith(steps.commit_message.outputs.commit_message, 'Version change')
      - uses: csexton/create-release@add-body
        id: create_release
        with:
          tag_name: ${{ steps.versioning.outputs.cli_version }}
          release_name: safe-cli
          draft: false
          prerelease: false
          body: ${{ steps.release_description.outputs.description }}
</code></pre>
<p>This works without any errors. However, the problem is, the description (the body parameter of the create_release action) is basically a small markdown document, so it’s a multiline string. The problem is that ::set-output is truncating everything except the first line. I can confirm that by echoing the description in the same action where it’s assigned and then referencing the description output variable in the next action.</p>
<p>Could someone please advise how I can get this to work with a multiline string?</p>
<p>Thanks. </p>

**Comment by Yanjingzhu** [comment]
<p>I have reported your question to the appropriate engineering team for further evaluation. The team will review the feedback and notify me about the next steps. I will update here in time. Thank you for your understanding.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>I got response from the team. </p>
<p>% and \n and \r can be escaped like below, the runner will unescape in reverse.</p>
<pre><code>content="${content//'%'/'%25'}"
content="${content//$'\n'/'%0A'}"
content="${content//$'\r'/'%0D'}"
</code></pre>
<p>Please try to add next three red lines to your yml, kindly let me know whether this could help.</p>
<pre><code>- shell: bash
        id: release_description
        run: |
          description=$(./resources/get_release_description.sh ${{ steps.versioning.outputs.cli_version }})
          echo $description
&lt;font color="#FF0000"&gt; description="${description//'%'/'%25'}"
          description="${description//$'\n'/'%0A'}"
          description="${description//$'\r'/'%0D'}"&lt;/font&gt;
echo "::set-output name=description::$description"
if: startsWith(steps.commit_message.outputs.commit_message, 'Version change')
</code></pre>
<p> I tested in my side, after adding these lines, I can use echo <strong>"</strong> ${{ steps.release_description.outputs.description1 }} <strong>"</strong> to output multiple line value. Please pay attention to <strong>“”</strong>. </p>
<p><img src="https://user-images.githubusercontent.com/24806478/181096678-c28d60cb-f681-4940-b1bc-9d9bcf09ff23.png" alt="multiline.png" data-base62-sha1="eo97k61cIRYOesNE3KyvBjwjHLY" width="524" height="411"></p>

**Comment by jacderida** [comment]
<p>Sorry, I just noticed your reply!</p>
<p>Thanks very much for this, it resolves the issue.</p>

**Comment by andreasplesch** [comment]
<p>It took me an hour to realize that multi-line strings were the issue for me, and then a few more hours to find this solution.</p>
<p>It would be great if multi-line strings as values could be better supported, and documented.</p>
<p>Here are a few observation. First, it is important to suppress word-splitting upon expansion in bash. This is easiest done by enclosing with double quotes:</p>
<pre><code>REPORT="$(cat logfile)"
</code></pre>
<p>Similarly, word splitting also has to be suppressed when reading back:</p>
<pre><code>echo "$REPORT"
</code></pre>
<p>The double quotes are necessary.</p>
<p>However, Workflows gets confused/truncates when dealing with an expanded multi-line value:</p>
<pre><code>echo "::set-env name=REPORT::$REPORT"
</code></pre>
<p>This does not work for multi-line variables.</p>
<p>The solution above uses a standard bash feature to expand and substitute characters:</p>
<pre><code>REPORT="${REPORT//'%'/'%25'}"
REPORT="${REPORT//$'\n'/'%0A'}"
REPORT="${REPORT//$'\r'/'%0D'}"
</code></pre>
<p>This essentially makes the value into a single line string.</p>
<p>The github engineers then also knew that when reading back the variable in the workflow these escape values get substituted back to the actual characters:</p>
<pre><code>${{ env.REPORT }}
</code></pre>
<p>This results actually in a multi-line string. Unfortunately, this is not documented anywhere and comes across a bit as magic. If there are other substitutions it would be good to also document those.</p>

**Comment by andreasplesch** [comment]
<p>Thanks so much posting this. I came across the same issue and found this rather obscure solution, after a lot of trial and error.</p>
<p>A couple observation dealing with multiline values, mostly as a reminder on how this works.</p>
<p>If using bash, it is important to suppress word splitting upon parameter expansion, to keep the line breaks. It is easiest with double quotes:</p>
<pre><code>REPORT="$(cat log.out)"
</code></pre>
<p>Accessing the variable also invokes word splitting, which needs to be suppressed:</p>
<pre><code>echo "$REPORT"
</code></pre>
<p>However, set-output or set-env does not work with expanded multiline values:</p>
<pre><code>echo "::set-env name=REPORT::$REPORT"
</code></pre>
<p>This does does not work, as Workflows somehow truncates or ignores the value.</p>
<p>The solution above is to escape the newlines and other apparently unhandled characters using bash expansion and substitution. This makes the value effectively single line.</p>
<p>The thing to notice is that Workflows substitutes the escaped characters back when the parameters is used in ${{ }}:</p>
<pre><code>body: |
  eslint and diff report:
  ```
  ${{ env.REPORT }}
  ```
</code></pre>
<p>Here the final value includes actual newlines.</p>
<p>It would be great if this behaviour cold be documented somewhere. For example, are there other substitutions happening ?</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/andreasplesch">@andreasplesch</a>  Thank you for your further investigation. I would recommand you to create an issue <a href="https://github.com/actions/toolkit/issues" rel="nofollow noopener">in the action toolkit repo</a> . You could ask for adding an example for <strong>set-env</strong> and <strong>set-output</strong> with multiline values in <a href="https://github.com/actions/toolkit/blob/master/docs/commands.md" rel="nofollow noopener">this</a> document. </p>

**Comment by andreasplesch** [comment]
<p>Thanks. Good to know that there is more in depth documentation available in the toolkit repo. I think a link from the main help page at <a href="https://help.github.com/en/actions" rel="nofollow noopener">https://help.github.com/en/actions</a> to this documentation could be helpful.</p>
<p>Hm, it looks like newlines should already be escaped automatically:</p>
<p><a href="https://github.com/actions/toolkit/blob/master/packages/core/src/command.ts#L76" rel="nofollow noopener">https://github.com/actions/toolkit/blob/master/packages/core/src/command.ts#L76</a></p>
<p>Anyways, here is the new issue in the repo:</p>
<p><a href="https://github.com/actions/toolkit/issues/403" rel="nofollow noopener">https://github.com/actions/toolkit/issues/403</a></p>
<p>and a related issue:</p>
<p><a href="https://github.com/actions/toolkit/issues/193" rel="nofollow noopener">https://github.com/actions/toolkit/issues/193</a></p>

**Comment by andreasplesch** [comment]
<p>Escaping of special characters already happened for javascript actions but does not for bash scripts. There is a plan to update documentation.</p>

**Comment by bewuethr** [comment]
I have bumped into this exact issue, and found another solution: we can use `jq` to do the conversion into a valid JSON string for us, and then use `fromJSON` to unpack it again.

For example: the variable `$mystr` holds the (multiline) contents of file `myfile`; we convert to a one-line JSON string with `jq --raw-input --slurp '.'`, and then unpack in a later step.

```yaml
- id: mystep
  run: |
    mystr=$(&lt; myfile)
    msg=$(printf '%s' "$mystr" | jq --raw-input --slurp '.')
    echo "::set-output name=msg::$msg"

- run: |
    echo "${{ fromJSON(steps.mystep.outputs.msg) }}"
```

Notice that this isn’t fool-proof, either: if the expanded string contains double quotes, the `echo` in the second step might get confused. Set the value as an environment variable to avoid.

**Comment by AraHaan** [comment]
<p>I also have an issue where I would like to output a list of paths and other data only if the action processes multiple files (to push them to <a href="http://nuget.org" rel="noopener nofollow ugc">nuget.org</a> or any nuget feeds).</p>

**Comment by pestophagous** [comment]
<p>Here is yet another potential work-around.</p>
<p>This is working in a CI Action job of mine, which takes the content of a text file that contains test results and posts the whole content (which is multiline content) back to the relevant PR as a comment. The resulting comment accurately reflects the original newlines of the test result output.</p>
<pre><code class="lang-auto">name: build_for_PR

on:
  pull_request:
    branches:
      - main

jobs:
  linux_publish_test_output:

    runs-on: ubuntu-18.04

    steps:
    - uses: actions/checkout@v1

    - name: run_all_linux_ci
      run: ./linux_build_all_and_test.sh

    - name: get_legacy_gui_test_output
      id: legacy_out_01
      run: echo "::set-output name=testconfout::$(cat path_from_my_git_root/to/build/artifacts/db_mgr_test_output.txt | python3 -c 'import sys; print(sys.stdin.read().replace("\\","\\\\").replace("\"","\\\"").replace("\n","\\n"))' )\n"

    - uses: actions/github-script@v3
      with:
        script: |
          var theContent = "${{join(steps.legacy_out_01.outputs.*, '\n')}}";
          github.issues.createComment({
          issue_number: context.issue.number,
          owner: context.repo.owner,
          repo: context.repo.repo,
          body: 'LEGACY GUI TEST OUTPUT (from most recent CI job on this PR):\n\n' + theContent
          })

</code></pre>
<p>This was inspired by the approach posted by @ <a href="https://github.community/u/bewuethr">bewuethr</a>. As far as I can tell in a couple days of usage, the python-based solution I am sharing does manage to handle double-quotes in the text (in addition to newlines).</p>

**Comment by PedroMartinSteenstrup** [comment]
<p>Thanks a lot for sharing, this unblocked me :slight_smile:</p>

**Comment by Simran-B** [comment]
<p>For completeness, I want to point out that you can also use an environment variable via environment files for multiline text: <a href="https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#multiline-strings" class="inline-onebox" rel="noopener nofollow ugc">Workflow commands for GitHub Actions - GitHub Docs</a></p>
<pre><code class="lang-yaml">steps:
  - name: Set the value
    shell: bash
    run: |
      echo 'JSON_RESPONSE&lt;&lt;EOF' &gt;&gt; $GITHUB_ENV
      curl https://httpbin.org/json &gt;&gt; $GITHUB_ENV
      echo 'EOF' &gt;&gt; $GITHUB_ENV
  - name: Print multiline response
    shell: bash
    run: |
      echo "$JSON_RESPONSE"
</code></pre>

**Reply by John4650-hub** [reply]
Thanks alot, this worked for me.

**Comment by Starlight220** [comment]
<p>Has this been fixed? I built a function to work around this, and now it displays the escape sequences.</p>

**Comment by MolotovCherry** [comment]
<p>Here’s my multi-line output solution which requires no changing of the data, however you do have to write it to a text file first.</p>
<p>In this example I’m using it for stderr, but you can adapt it to any kind of multi-line output too</p>
<pre><code class="lang-auto"># on windows, powershell fails on stderr output and improperly
# handles putting it in the file. It's required to run as cmd /c
# if not using windows, then change this command appropriately
- name: build
  id: build
  continue-on-error: true
  run: cmd /c "build 2&gt; errors.txt"

- name: Set error-log to var
  uses: actions/github-script@v4
  id: error-log
  if: steps.build.outcome != 'success'
  with:
    script: |
      const fs = require('fs');
      return fs.readFileSync('errors.txt','utf8').toString();
    result-encoding: string

# you can access the output variable in like so
- name: Some action
  uses: whateveryouwant/whatever@4
  if: steps.build.outcome != 'success'
  with:
    something: ${{ steps.error-log.outputs.result }}

# print errors.txt to stdout + fail workflow since build failed
# use `cat` if not using windows
- name: Show build errors
  if: steps.build.outcome != 'success'
  run: |
    type errors.txt
    exit 1
</code></pre>

**Comment by MolotovCherry** [comment]
<p>Note: You can change the output variable name by using <code>core.setOutput</code> as well.</p>

**Comment by mirrorcult** [comment]
<p>For any lost souls in the future: If you’re setting output in a Docker container (like if your action is a .NET binary), not in the action YML, by just printing <code>::set-output name=whatever::whatever</code>, all you need to do to fix this is replace any <code>\n</code> you have with <code>%0A</code></p>

**Comment by elb-slack** [comment]
<p>Also for anyone in the future with this issue, I JSON-encoded the content I needed to put into <code>GITHUB_ENV</code> like so:</p>
<pre><code class="lang-auto">cat_results=$(cat results/results_blob.txt | python3 -c 'import json; import sys; print(json.dumps(sys.stdin.read()))')
echo "RESULTS=${cat_results}" &gt;&gt; $GITHUB_ENV
</code></pre>
<p>I kept getting errors with the multiline <code>$GITHUB_ENV</code>. This worked perfectly since I wanted the content JSON-encoded anyway. It was sent over the Github API to create a comment on a PR. You can decode the JSON in another step rather than hoping you caught all url-encoded characters with <code>replace()</code>.</p>

**Comment by bdovh** [comment]
Wow, I had this problem for a year and I couldn't solve it. I believed this one the problem with slack action that wasn't supporting multiline comments. Now I have finally found this thread and found a solution for a problem in my script that existed more than a year

**Comment by ffried** [comment]
For future reference:
With [GitHub's new `GITHUB_OUTPUT` file](https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/), multiline outputs don't need to be escaped anymore. They simply [need to use a delimiter](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings):

```bash
delimiter="$(openssl rand -hex 8)"
echo "output-name<<${delimiter}" >> "${GITHUB_OUTPUT}"
echo "Some\nMultiline\nOutput" >> "${GITHUB_OUTPUT}"
echo "${delimiter}" >> "${GITHUB_OUTPUT}"
```

Note I'm using `openssl rand -hex 8` to get a unique delimiter as suggested in the docs.

**Reply by lukasz-mitka** [reply]
Did you see any issues with simple `EOF` delimiter? It's used in the example btw.

**Reply by ffried** [reply]
Didn't try it to be honest. The warning states that it should be random and unique.
Btw. `@actions/core`'s `setOutput` also [uses a generated identifier](https://github.com/actions/toolkit/blob/ffb7e3e14ed5e28ae00e9c49ba02b2764d57a6b7/packages/core/src/file-command.ts#L28).

**Reply by lukasz-mitka** [reply]
But content unique, locally unique, job unique, workflow unique, globally unique?
I wonder since rand generated makes it more complex and wonder if it's necessary.
Like will `EOF` fail with multiple multiline outputs 🤷
Guess I will check one day :)

**Reply by ffried** [reply]
The reason is trying to prevent [script injection](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections).
`EOF` will probably work just fine, but might pose a security issue. I opted for using a uniquely generated delimiter because it's basically just one more line of code - thus an easy security win.

**Reply by lukasz-mitka** [reply]
Thank you! I managed to miss that part 🤦

**Reply by norwd** [reply]
Thanks for posting! Accepted answer didn't work for me (albeit I have a weird workflow file), but this method worked perfectly!

**Reply by manumaan** [reply]
You are a proper lifesaver cos this helped me with my multi-line output

**Reply by karfau** [reply]
It would be really nice to add those instractions to https://docs.github.com/en/actions/using-jobs/defining-outputs-for-jobs

**Reply by bewuethr** [reply]
It's documented in [workflow commands](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings).

**Reply by plocket** [reply]
When I used this in a composite action and created an output with it, my mutliline string lost all of its new lines. When I tried to echo it later, it was all on one line. How do I preserve the new lines? Or did I do something wrong?

In the composite action it echos as a multiline string:
```yml
outputs:
  TEXT:
    description: 'Text from action'
    value: ${{ steps.text_output.outputs.TEXT }}
# Lower down in a step in a job
id: text_output
run: |
  echo "$TEXT"
  echo 'TEXT<<EOF' >> $GITHUB_OUTPUT
  echo $TEXT >> $GITHUB_OUTPUT
  echo 'EOF' >> $GITHUB_OUTPUT
```

In the workflow that uses the composite action, it echos the right text, but all on a single line:
```yml
run: |
  echo "TEXT=${{ steps.the_text.outputs.TEXT }}" >> $GITHUB_ENV
```

When I put it in the body of an issue using the GitHub cli, it also appears as just one line of text.

**Reply by ffried** [reply]
This is probably because you don't quote the `$TEXT` variable when passing it to `echo` the second time. This will lead to multiple arguments passed to `echo` which in turn looses the newlines.

Try using `echo "$TEXT" >> "$GITHUB_OUTPUT"` instead. 
(btw. it's generally good practice to "quote everything" in shell scripts - with a few specific exceptions of course).

**Reply by bewuethr** [reply]
See this Stack Overflow Q&A: [*I just assigned a variable, but echo $variable shows something else*](https://stackoverflow.com/q/29378566/3266847)

Else, any time you have multi-line values you want to set in the environment or as an output, you *have* to use the `<<EOF` style syntax in Actions.
